flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/strategy/AdaptiveSkewedJoinOptimizationStrategy.java

```diff
@@ -0,0 +1,317 @@
+                        .get(adaptiveJoinNode.getId())
+                        .get(LEFT_INPUT_TYPE_NUMBER);
+        long[] rightInputSize =
+                aggregatedProducedBytesByTypeNumberAndNodeId
+                        .get(adaptiveJoinNode.getId())
+                        .get(RIGHT_INPUT_TYPE_NUMBER);
+        checkArgument(
+                leftInputSize != null && rightInputSize != null,
+                "Adaptive join node currently supports only two inputs, "
+                        + "but received input bytes with left [%s] and right [%s] for stream "
+                        + "node id [%s].",
+                leftInputSize,
+                rightInputSize,
+                adaptiveJoinNode.getId());
+
+        long leftSkewedThreshold =
+                computeSkewThreshold(median(leftInputSize), skewedFactor, skewedThresholdInBytes);
+        long rightSkewedThreshold =
+                computeSkewThreshold(median(rightInputSize), skewedFactor, skewedThresholdInBytes);
+
+        boolean isLeftOptimizable = false;
+        boolean isRightOptimizable = false;
+        switch (joinType) {
+            case RIGHT:
+                isRightOptimizable = true;
+                break;
+            case INNER:
+                isLeftOptimizable = true;
+                isRightOptimizable = true;
+                break;
+            case LEFT:
+            case SEMI:
+            case ANTI:
+                isLeftOptimizable = true;
+                break;
+            case FULL:
+            default:
+                throw new RuntimeException(String.format("Unexpected join type %s.", joinType));
+        }
+
+        isLeftOptimizable =
+                isLeftOptimizable
+                        & existBytesLargerThanThreshold(leftInputSize, leftSkewedThreshold);
+        isRightOptimizable =
+                isRightOptimizable
+                        & existBytesLargerThanThreshold(rightInputSize, rightSkewedThreshold);
+
+        if (isLeftOptimizable) {
+            boolean isModificationSucceed =
+                    postStreamEdgeModificationRequest(
+                            context, adaptiveJoinNode, LEFT_INPUT_TYPE_NUMBER);
+            LOG.info(
+                    "Apply skewed join optimization {} for left input of node {}.",
+                    isModificationSucceed ? "succeeded" : "failed",
+                    adaptiveJoinNode.getId());
+        }
+        if (isRightOptimizable) {
+            boolean isModificationSucceed =
+                    postStreamEdgeModificationRequest(
+                            context, adaptiveJoinNode, RIGHT_INPUT_TYPE_NUMBER);
+            LOG.info(
+                    "Apply skewed join optimization {} for right input of node {}.",
+                    isModificationSucceed ? "succeeded" : "failed",
+                    adaptiveJoinNode.getId());
+        }
+    }
+
+    private static boolean postStreamEdgeModificationRequest(
```

Comment:
maybe `tryModifyInput` and add comments to describe all the modifications it does.

`post` can be ambiguous here.